docs/repo/crates/discv4.md

```diff
@@ -0,0 +1,286 @@
+            if self.config.enable_lookup
+                && !self.is_lookup_in_progress()
+                && self.lookup_interval.poll_tick(cx).is_ready()
+            {
+                let target = self.lookup_rotator.next(&self.local_node_record.id);
+                self.lookup_with(target, None);
+            }
+
+            // evict expired nodes
+            if self.evict_expired_requests_interval.poll_tick(cx).is_ready() {
+                self.evict_expired_requests(Instant::now())
+            }
+
+            // re-ping some peers
+            if self.ping_interval.poll_tick(cx).is_ready() {
+                self.re_ping_oldest();
+            }
+
+            if let Some(Poll::Ready(Some(ip))) =
+                self.resolve_external_ip_interval.as_mut().map(|r| r.poll_tick(cx))
+            {
+                self.set_external_ip_addr(ip);
+            }
+        // --snip--
+        }
+    }
+}
+```
+
+As the function starts, a `loop` is entered and the `Discv4Service.queued_events` are evaluated to see if there are any events ready to be processed. If there is an event ready, the function immediately returns the event wrapped in `Poll::Ready()`. The `queued_events` field is a `VecDeque<Discv4Event>` where `Discv4Event` is an enum containing one of the following variants.
+
+[File: ]()
+```rust ignore
+pub enum Discv4Event {
+    /// A `Ping` message was handled.
+    Ping,
+    /// A `Pong` message was handled.
+    Pong,
+    /// A `FindNode` message was handled.
+    FindNode,
+    /// A `Neighbours` message was handled.
+    Neighbours,
+    /// A `EnrRequest` message was handled.
+    EnrRequest,
+    /// A `EnrResponse` message was handled.
+    EnrResponse,
+}
+```
+
+If there is not a `Discv4Event` immediately ready, the function continues triggering self lookup, and removing any nodes that have not sent a ping request in the allowable time elapsed (`now.duration_since(enr_request.sent_at) < Discv4Service.config.ping_expiration`). Additionally, the node re-pings all nodes whose endpoint proofs are considered expired. If the node fails to respond to the "ping" with a "pong", the node will be removed from the `KbucketsTable`.
+
+To prevent traffic amplification attacks (ie. DNS attacks), implementations must verify that the sender of a query participates in the discovery protocol. The sender of a packet is considered verified if it has sent a valid Pong response with matching ping hash within the last 12 hours.
+
+Next, the Discv4Service handles all incoming [`Discv4Command`](https://github.com/paradigmxyz/reth/blob/main/crates/net/discv4/src/lib.rs#L1527)s until there are no more commands to be processed. Following this, All `IngressEvent`s are handled, which represent all incoming datagrams related to the discv4 protocol. After all events are handled, the node pings to active nodes in it's network. This process is repeated until all of the `Discv4Event`s in `queued_events` are processed. 
```

Comment:
```suggestion
Next, the Discv4Service handles all incoming `Discv4Command`s until there are no more commands to be processed. Following this, All `IngressEvent`s are handled, which represent all incoming datagrams related to the discv4 protocol. After all events are handled, the node pings to active nodes in it's network. This process is repeated until all of the `Discv4Event`s in `queued_events` are processed. 
```

Again, wary of links to code since structure might change, also you show what the `Discv4Event` enum looks like above